### PR TITLE
[Uid] Improve the code

### PR DESCRIPTION
**Pull request by nilmarsm86**
Improve the reuse of code, programing for extension and not for modification, create a Trait with similar behavior for getTime method in uuid types. Eliminate duplicate code in Uuid

| Q             | A
| ------------- | ---
| Branch?       | 5.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #...
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

Keep the same functionality, just improve a little the code, eliminating duplications and reusing some lines.

### CODE REVIEW

File: src/Symfony/Component/Uid/Uuid.php [hunk 1]
```diff
@@ -133,4 +127,12 @@ public function compare(parent $other): int
 
         return parent::compare($other);
     }
+
+    private static function encodingUuid(string $uuid, $start): string
```
[the remark this answers is not in the document]
Comment (nilmarsm86) [reply src/Symfony/Component/Uid/Uuid.php]: OK, thanks in minutes

File: src/Symfony/Component/Uid/UuidTrait.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+namespace Symfony\Component\Uid;
+
+/**
+ * @experimental in 5.1
+ *
+ * @author Grégoire Pineau <lyrixx@lyrixx.info>
+ */
+trait UuidTrait
```
[the remark this answers is not in the document]
Comment (nilmarsm86) [reply src/Symfony/Component/Uid/UuidTrait.php]: Ok, thanks, for that reason you are the rock!!!!!

**Comment by nilmarsm86** [comment]
I am sorry it is my first time doing this. Fixed the recommendation.

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Uid/Uuid.php [hunk 3]
```diff
@@ -76,10 +76,7 @@ public static function fromString(string $uuid): parent
     {
         // don't use uuid_generate_md5(), some versions are buggy
         $uuid = md5(hex2bin(str_replace('-', '', $namespace->uid)).$name, true);
-        $uuid[8] = $uuid[8] & "\x3F" | "\x80";
-        $uuid = substr_replace(bin2hex($uuid), '-', 8, 0);
-        $uuid = substr_replace($uuid, '-3', 13, 1);
-        $uuid = substr_replace($uuid, '-', 18, 0);
+        $uuid = static::format($uuid, '-3');
 
         return new UuidV3(substr_replace($uuid, '-', 23, 0));
```
Comment (rosier) [inline src/Symfony/Component/Uid/Uuid.php]: This `substr_replace` could be moved to `static::format` also

File: src/Symfony/Component/Uid/Uuid.php [hunk 4]
```diff
@@ -93,10 +90,7 @@ public static function fromString(string $uuid): parent
     {
         // don't use uuid_generate_sha1(), some versions are buggy
         $uuid = substr(sha1(hex2bin(str_replace('-', '', $namespace->uid)).$name, true), 0, 16);
-        $uuid[8] = $uuid[8] & "\x3F" | "\x80";
-        $uuid = substr_replace(bin2hex($uuid), '-', 8, 0);
-        $uuid = substr_replace($uuid, '-5', 13, 1);
-        $uuid = substr_replace($uuid, '-', 18, 0);
+        $uuid = static::format($uuid, '-5');
 
         return new UuidV5(substr_replace($uuid, '-', 23, 0));
```
Comment (rosier) [inline src/Symfony/Component/Uid/Uuid.php]: Same here

File: src/Symfony/Component/Uid/Uuid.php (hunk 3 above)
Comment (nilmarsm86) [reply src/Symfony/Component/Uid/Uuid.php]: ok, thanks

File: src/Symfony/Component/Uid/Uuid.php (hunk 4 above)
Comment (nilmarsm86) [reply src/Symfony/Component/Uid/Uuid.php]: ok, thanks

**Review by rosier — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @nilmarsm86.

**Pull request merged**